### PR TITLE
sql: add `switchToAnotherPortal` signal for result consumer

### PR DESCRIPTION
**Pull request by ZhouXing19**
This PR is part of the effort to implement the multiple active portals. (Extracted from #96358)

---

### sql/settings: add sql.pgwire.multiple_active_portals.enabled cluster setting

This commit is to add a non-public `sql.pgwire.multiple_active_portals.enabled`
setting. This setting is only for a PREVIEWABLE feature.
With it set to true, all non-internal portals with read-only queries without sub/post
queries can be paused and resumed in an interleaving manner but are executed with
local plan.

---
### sql: add switchToAnotherPortal signal for result consumer

Previously, after pushing the result to the consumer, we only accept the
following actions as the next step:
1. Pushing more data from the same source to the same consumer;
2. Drain or close the pipeline.

This commit is to add another option: pause the current execution, and switch
to the execution to another portal. I.e. when receiving an `ExecPortal` cmd but
for another portal, we do nothing and return the control to the connExecutor.
This allows us to execute different portals interleaving-ly.

Epic: CRDB-17622

Release note (sql change): add a non-public `sql.pgwire.multiple_active_portals.enabled` setting. This setting is only for a PREVIEWABLE feature. With it set to true, all non-internal portals with read-only queries without sub/post queries can be paused and resumed in an interleaving manner but are executed with local plan.

### CODE REVIEW

**Comment by rytaft** [comment]
I think you want @yuzefovich to review this, not me.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r1, 12 of 12 files at r3, 11 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-:-NR-c-cA0gCV9gu5KpsX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`pkg/sql/prepared_stmt.go` line 135 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-NR-bEG83RoRg4hFlqrX:-NR-bEG83RoRg4hFlqrY:bwfv18d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a162fbf7d367458aa2872312ca15ac73806c56/pkg/sql/prepared_stmt.go#L135)):*
> ```Go
> 	// NotPausablePortalForUnsupportedStmt is used when the cluster setting
> 	// sql.defaults.multiple_active_portals.enabled is set to true, while we don't
> 	// support underlying statement it.
> ```

nit: something is off in "underlying statement it".
___
*[`pkg/sql/execinfra/base.go` line 205 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-NR-bWtv9Lq4TG1e6XtB:-NR-bWtv9Lq4TG1e6XtC:b-nd4hnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a162fbf7d367458aa2872312ca15ac73806c56/pkg/sql/execinfra/base.go#L205)):*
> ```Go
> 				dst.ProducerDone()
> 				return
> 			case SwitchToAnotherPortal:
> ```

nit: perhaps move `SwitchToAnotherPortal` case right after `NeedMoreRows` here and in `DrainAndForwardMetadata`.
___
*[`pkg/sql/pgwire/command_result.go` line 494 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-NR-bm0q-AXkZfq0dvet:-NR-bm0q-AXkZfq0dveu:bx6gvks) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a162fbf7d367458aa2872312ca15ac73806c56/pkg/sql/pgwire/command_result.go#L494)):*
> ```Go
> // when a limit has been specified.
> func (r *limitedCommandResult) moreResultsNeeded(ctx context.Context) error {
> 	// Keep track of the previous CmdPos so we can rewind if needed.
> ```

nit: move this comment below to be on top of `prevPos := ...` line.

**Review by ZhouXing19 — COMMENTED** [review]
TFTR!
cc @rafiss I think we can leave the cluster setting `false` by default for now, and enable the metamorphic setting after the whole functionality is merged? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-:-NR-dwgs1kEYcPMdxRZS:bjaal1h)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

File: pkg/sql/exec_util.go [hunk 1]
```diff
@@ -733,6 +733,14 @@ var overrideAlterPrimaryRegionInSuperRegion = settings.RegisterBoolSetting(
 	false,
 ).WithPublic()
 
+var enableMultipleActivePortals = settings.RegisterBoolSetting(
+	settings.TenantWritable,
+	"sql.defaults.multiple_active_portals.enabled",
```
Comment (rafiss) [inline pkg/sql/exec_util.go:738]: nit: sql.defaults settings are for session variable defaults. instead, can we name this sql.pgwire.multiple_active_portals.enabled? (be sure to update the release note)

and i think this should be public

**Review by rafiss — CHANGES_REQUESTED** [review]
copying some comments from #96358

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99052)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich and @ZhouXing19)

___
*[`pkg/sql/distsql_running.go` line 1517 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-NR-mbH-9Qt1ykcNI6qy:-NR-mbH-9Qt1ykcNI6qz:b1u0pcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a83f318841c815a57cc4373da920901aeb93a756/pkg/sql/distsql_running.go#L1517)):*
> ```Go
> 		"multiple active portals not supported, "+
> 			"please set sql.pgwire.multiple_active_portals.enabled to true. "+
> 			"(Note this feature is in preview)",
> ```

nit: it should say `Note: this feature is in preview"

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99052)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich and @ZhouXing19)

___
*[`-- commits` line 4 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-NR-mpHW-Lzaaj8LFAAw):*
nit: it is public now

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99052)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich and @ZhouXing19)

___
*[`pkg/sql/exec_util.go` line 742 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99052#-NR-mtYb04OZnHadt1SC:-NR-mtYb04OZnHadt1SD:bn65r8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a83f318841c815a57cc4373da920901aeb93a756/pkg/sql/exec_util.go#L742)):*
> ```Go
> 		"can be executed in interleaving manner, but with local execution plan",
> 	false,
> )
> ```

let's make it public. we still want people to see the setting, but they just need to opt-in to it. (this requires `./dev gen docs` too)

**Comment by ZhouXing19** [comment]
TFTR!
bors r+

**Pull request merged**